Mathlib/Data/Finset/Density.lean

```diff
@@ -0,0 +1,166 @@
+
+@[simp] lemma dens_singleton (a : α) : dens ({a} : Finset α) = (Fintype.card α : ℚ≥0)⁻¹ := by
+  simp [dens]
+
+@[simp] lemma dens_cons (h : a ∉ s) : (s.cons a h).dens = dens s + (Fintype.card α : ℚ≥0)⁻¹ := by
+  simp [dens, add_div]
+
+@[simp] lemma dens_disjUnion (s t : Finset α) (h) : dens (s.disjUnion t h) = dens s + dens t := by
+  simp_rw [dens, card_disjUnion, Nat.cast_add, add_div]
+
+@[simp] lemma dens_eq_zero : dens s = 0 ↔ s = ∅ := by
+  simp (config := { contextual := true }) [dens, Fintype.card_eq_zero_iff, eq_empty_of_isEmpty]
+
+lemma dens_ne_zero : dens s ≠ 0 ↔ s.Nonempty := dens_eq_zero.not.trans nonempty_iff_ne_empty.symm
+
+@[simp] lemma dens_pos : 0 < dens s ↔ s.Nonempty := pos_iff_ne_zero.trans dens_ne_zero
+
+protected alias ⟨_, Nonempty.dens_pos⟩ := dens_pos
+protected alias ⟨_, Nonempty.dens_ne_zero⟩ := dens_ne_zero
+
+lemma dens_le_dens (h : s ⊆ t) : dens s ≤ dens t :=
+  div_le_div_of_nonneg_right (mod_cast card_mono h) $ by positivity
+
+lemma dens_lt_dens (h : s ⊂ t) : dens s < dens t :=
+  div_lt_div_of_pos_right (mod_cast card_strictMono h) $ by
+    cases isEmpty_or_nonempty α
+    · simp [Subsingleton.elim s t, ssubset_irrfl] at h
+    · exact mod_cast Fintype.card_pos
+
+@[mono] lemma dens_mono : Monotone (dens : Finset α → ℚ≥0) := fun _ _ ↦ dens_le_dens
+@[mono] lemma dens_strictMono : StrictMono (dens : Finset α → ℚ≥0) := fun _ _ ↦ dens_lt_dens
+
+lemma dens_map_le [Fintype β] (f : α ↪ β) : dens (s.map f) ≤ dens s := by
+  cases isEmpty_or_nonempty α
+  · simp [Subsingleton.elim s ∅]
+  simp_rw [dens, card_map]
+  gcongr
+  · positivity
+  · exact mod_cast Fintype.card_pos
+  · exact Fintype.card_le_of_injective _ f.2
+
+section Nonempty
+variable [Nonempty α]
+
+@[simp] lemma dens_univ : dens (univ : Finset α) = 1 := by simp [dens, card_univ]
+
+@[simp] lemma dens_eq_one : dens s = 1 ↔ s = univ := by
+  simp [dens, div_eq_one_iff_eq, card_eq_iff_eq_univ]
+
+lemma dens_ne_one : dens s ≠ 1 ↔ s ≠ univ := dens_eq_one.not
+
+end Nonempty
+
+section Lattice
+variable [DecidableEq α]
+
+lemma dens_union_add_dens_inter (s t : Finset α) :
+    dens (s ∪ t) + dens (s ∩ t) = dens s + dens t := by
+  simp_rw [dens, ← add_div, ← Nat.cast_add, card_union_add_card_inter]
+
+lemma dens_inter_add_dens_union (s t : Finset α) :
+    dens (s ∩ t) + dens (s ∪ t) = dens s + dens t := by rw [add_comm, dens_union_add_dens_inter]
+
+@[simp] lemma dens_union_of_disjoint (h : Disjoint s t) : dens (s ∪ t) = dens s + dens t := by
+  rw [← disjUnion_eq_union s t h, dens_disjUnion _ _ _]
+
+lemma dens_sdiff_add_dens_eq_dens (h : s ⊆ t) : dens (t \ s) + dens s = dens t := by
+  simp [dens, ← card_sdiff_add_card_eq_card h, add_div]
+
+lemma dens_sdiff_add_dens (s t : Finset α) : dens (s \ t) + dens t = (s ∪ t).dens := by
+  rw [← dens_union_of_disjoint sdiff_disjoint, sdiff_union_self_eq_union]
+
+lemma dens_sdiff_comm [IsCancelAdd ℚ≥0] (h : card s = card t) : dens (s \ t) = dens (t \ s) :=
```

Comment:
I guess it can work for prop typeclasses, but the `IsCancelAdd ℚ≥0` argument looks pretty suspicious to me

Reply:
Whoops, that's a remnant from the previous approach